presto-main/src/main/java/io/prestosql/operator/unnest/BlockProducerFromSource.java

```diff
@@ -0,0 +1,337 @@
+            }
+        }
+        else {
+            for (int i = 0; i < length; i++) {
+                appendId(startPosition + i);
+            }
+        }
+    }
+
+    /**
+     * Prepare for appending the element at position {@code index} in the source
+     * for {@code count} number of times in the output
+     *
+     * Purpose of this method is to avoid repeated checks (for the same
+     * element) in {@link #appendElementFromSource}.
+     * @param index
+     * @param count
+     */
+    public void appendRepeated(int index, int count)
+    {
+        checkIndexValid(index);
+
+        if (usingCopiedBlock) {
+            for (int i = 0; i < count; i++) {
+                type.appendTo(source, index, outputBlockBuilder);
+            }
+        }
+        else {
+            for (int i = 0; i < count; i++) {
+                appendId(index);
+            }
+        }
+    }
+
+    private void checkIndexValid(int index)
+    {
+        checkState(source != null, "source is null");
+        checkArgument(index >= 0 && index <= source.getPositionCount());
+    }
+
+    private void checkRangeValid(int startIndex, int length)
+    {
+        checkState(source != null, "source is null");
+        checkArgument(startIndex >= 0, "startIndex is negative");
+        checkArgument(length >= 0, "length is negative");
+        checkArgument(startIndex + length - 1 < source.getPositionCount(), "range is not valid");
+    }
+
+    /**
+     * Prepare for appending null to output block
+     */
+    public void appendNull()
+    {
+        if (usingCopiedBlock) {
+            outputBlockBuilder.appendNull();
+        }
+        else {
+            int nullIndex = nullFinder.getNullElementIndex();
+            if (nullIndex == -1) {
+                startUsingCopiedBlock();
+                appendNull();
+            }
+            else {
+                appendId(nullIndex);
+            }
+        }
+    }
+
+    /**
+     * This method is invoked when transition from dictionary block
+     * to copied block happens.
+     *
+     * outputBlockBuilder is created and existing elements as indicated by
+     * ids[] are copied over.
+     *
+     * ids[] and positionCount are reset, as they are not used anymore.
+     */
+    private void startUsingCopiedBlock()
+    {
+        requireNonNull(pageBuilderStatus, "pageBuilderStatus is null");
+        // initialize state for copied block
+        outputBlockBuilder = type.createBlockBuilder(pageBuilderStatus.createBlockBuilderStatus(), ids.length);
+
+        // Transfer elements from ids[] to output page builder
+        for (int i = 0; i < positionCount; i++) {
+            type.appendTo(source, ids[i], outputBlockBuilder);
+        }
+
+        // reset state for dictionary block
+        ids = new int[0];
+        positionCount = 0;
+
+        usingCopiedBlock = true;
+    }
+
+    /**
+     * This method appends {@code sourcePosition} to ids array.
```

Comment:
If the source block is itself a dictionary, this will result in two or more levels of dictionary. It's best if we can collapse that. This may be an optimization that needs to be done after splitting this class in two.